### ISSUE TITLE
Why does imitation learning with demonstrations require Unity be running?

### ISSUE THREAD

**Issue by BartMartner** [Unity-Technologies/ml-agents#1532]
Hi! I was excited to see the addition of demonstrations in 0.6 but confused when I learned that training via demonstrations requires either a build or the Unity editor be running.

I understand the necessity of having a build running if your agents are using observations gathered from an in-game camera, but I would think you could simply back propagate using the pairs of observations and actions gathered in the demo otherwise.

Because it requires a build be running, imitation learning with demonstrations doesn't seem to save that much time over have real-time imitation learning.

For me, it'd be ideal if I could record multiple demos, and then have a python script back propagate the data from all of those recordings as fast as possible without the overhead of the editor or a build running.

Any plans for this?

**Comment by BartMartner** [comment]
Hmm... maybe that makes some sense, but I'm curious how that quality is evaluated. As I understand it there's currently no reward functions associated with ML Agent's implementation of BC. Is the agent's behavior compared to the demo somehow? I thought the demo was mostly just a record of inputs and outputs. My understanding of BC was that those inputs and outputs were just being repeatedly back-propagated using math. Does having the environment running verify something? What does it verify and how?

Typically when I've trained using offline imitation, the environment runs too slow to accurately recreated the desired behavior anyways. However, after the training the agent usually performs just fine.

For reference, I'm trying to get a platformer character to move to a desired location by jumping and moving left and right. To do that I've set up targets around various courses and when a target is reached by the player another random target is chosen. The player observes a grid of the environment (collision on or off), its own velocity, etc. I can train it to move around any given course, but it always seems to over fit to any specific course, becoming unable to run other courses.
https://www.youtube.com/watch?v=YPgocvJAknA

In an attempt to combat over-fitting, I'd like to have it train on something like a playlist of much shorter demos from across various courses (courses are scenes in this instance). I think it might be able to develop more general policies that way. Having have the editor running seems like a barrier to this.

Thanks for the reply!
